lib/SIL/SILBasicBlock.cpp

```diff
@@ -187,9 +187,32 @@ transferNodesFromList(llvm::ilist_traits<SILBasicBlock> &SrcTraits,
   // If splicing blocks not in the same function, update the parent pointers.
   for (; First != Last; ++First) {
     First->Parent = Parent;
-    for (auto &II : *First)
+    for (auto &II : *First) {
       II.setDebugScope(B,
                        ScopeCloner.getOrCreateClonedScope(II.getDebugScope()));
+      SILInstruction *I = &II;
+      // Maintain the registrations of opened archetypes.
+      if (isa<OpenExistentialAddrInst>(I) || isa<OpenExistentialRefInst>(I) ||
+          isa<OpenExistentialBoxInst>(I)) {
+        auto Ty = I->getType().getSwiftRValueType();
+        if (isa<ArchetypeType>(Ty) &&
+            !cast<ArchetypeType>(Ty)->getOpenedExistentialType().isNull()) {
+          SrcTraits.Parent->removeOpenedArchetypeDef(Ty, I);
+          Parent->addOpenedArchetypeDef(Ty, I);
+        }
+      } else if (isa<OpenExistentialMetatypeInst>(I)) {
+        SILType InstanceTy = I->getType();
+        while (isa<AnyMetatypeType>(InstanceTy.getSwiftRValueType())) {
+          InstanceTy = InstanceTy.getMetatypeInstanceType(I->getModule());
+        }
+        auto Ty = InstanceTy.getSwiftRValueType();
+        if (isa<ArchetypeType>(Ty) &&
+            !cast<ArchetypeType>(Ty)->getOpenedExistentialType().isNull()) {
+          SrcTraits.Parent->removeOpenedArchetypeDef(Ty, I);
+          Parent->addOpenedArchetypeDef(Ty, I);
+        }
+      }
+    }
```

Comment:
Regarding this and the similar changes in SILInstruction: I don't think we should make the rest of the compiler pay a cost for this one bit of functionality since it is easy to do this in the performance inliner and mandatory inliner (and without adding a map to Function).

We should just modify the inliners to do an RPO walk updating the map of opened types and providing that as an argument to the constructor of SILInliner.